vault/plugin_catalog.go

```diff
@@ -59,12 +84,98 @@ func (c *Core) setupPluginCatalog(ctx context.Context) error {
 	return nil
 }
 
+func (c *PluginCatalog) removeMultiplexedClient(ctx context.Context, name, id string) {
+	if _, ok := c.multiplexedClients[name]; !ok {
+		return
+	}
+
+	delete(c.multiplexedClients[name].connections, id)
+	c.logger.Debug("deleted multiplexedClients connection entry", "id", id)
+
+	if len(c.multiplexedClients[name].connections) == 0 {
+		// TODO(JM): This leaves child process behind after vault exits
+		c.multiplexedClients[name].client.Kill()
+		delete(c.multiplexedClients, name)
+		c.logger.Debug("deleted multiplexedClients plugin entry", "name", name)
+	}
+}
+
+func (c *PluginCatalog) getMultiplexedClient(pluginName string) *MultiplexedClient {
+	if mpc, ok := c.multiplexedClients[pluginName]; ok {
+		c.logger.Debug("MultiplexedClient exists", "pluginName", pluginName)
+
+		return mpc
+	}
+
+	c.logger.Debug("MultiplexedClient does not exist", "pluginName", pluginName)
+
+	return c.newMultiplexedClient(pluginName)
+}
+
+func (c *PluginCatalog) newMultiplexedClient(pluginName string) *MultiplexedClient {
+	if c.multiplexedClients == nil {
+		c.multiplexedClients = make(map[string]*MultiplexedClient)
+		c.logger.Debug("created multiplexedClients map")
+	}
+
+	mpc := &MultiplexedClient{
+		connections: make(map[string]plugin.ClientProtocol),
+	}
+
+	// set the MultiplexedClient for the given plugin name
+	c.multiplexedClients[pluginName] = mpc
+	c.logger.Debug("set the MultiplexedClient for", "pluginName", pluginName)
+
+	return mpc
+}
+
+// getPluginClient returns a client for managing the lifecycle of a plugin
+// process and an ID for a newly created entry in the MultiplexedClients'
+// connections map.
+func (c *PluginCatalog) getPluginClient(ctx context.Context, sys pluginutil.RunnerUtil, pluginRunner *pluginutil.PluginRunner, namedLogger log.Logger, isMetadataMode bool) (plugin.ClientProtocol, string, error) {
+	id, err := base62.Random(10)
+	if err != nil {
+		return nil, "", err
+	}
+
+	mpc := c.getMultiplexedClient(pluginRunner.Name)
+
+	if mpc.client == nil {
+		c.logger.Debug("spawning a new plugin process")
+		client, err := pluginRunner.RunConfig(ctx,
+			pluginutil.Runner(sys),
+			pluginutil.PluginSets(v5.PluginSets),
+			pluginutil.HandshakeConfig(v5.HandshakeConfig),
```

Comment:
These should probably be passed into this function instead of assuming we are generating a database plugin.

Perhaps a new arg type:
```
PluginClientConfig{
    Name string
    PluginSets ..
    HandshakeConfig ...
}
```

Reply:
Thanks! Yes, I had the same thought and I will consider the arg type

Reply:
addressed in: https://github.com/hashicorp/vault/pull/13734/commits/bee22a9ca5649da2dc01142094af018df8b430a3